### PR TITLE
Shopping cart us25

### LINKED ISSUE #31: User Story 25, Visitors must register or log in to check out
As a visitor
When I have items in my cart
And I visit my cart
I see information telling me I must register or log in to finish the checkout process
The word "register" is a link to the registration page
The words "log in" is a link to the login page

### PR DESCRIPTION
**Pull request by LHJE**
closes #31 

**Contributors**
- Driver: @LHJE 

**What was changed:**
- Adds tests to Checkout Spec to ensure there is a message prompting a visitor to login before finishing checking out.
- Adds said prompt to show page, but only for visitors.

**Any tech notes (if applicable)**
- I used  `!within` in the checkout_spec to designate that I wanted the test to click the `Register` button that's _not_ within the nav, and it appears to be working, but it's an almost un-google-able phrase so I'd like some eyes on the test on line 41 in that spec to see if it actually is doing what I believe it's doing.

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -13,13 +13,23 @@
 #merchants
 bike_shop = Merchant.create(name: "Meg's Bike Shop", address: '123 Bike Rd.', city: 'Denver', state: 'CO', zip: 80203)
 dog_shop = Merchant.create(name: "Brian's Dog Shop", address: '125 Doggo St.', city: 'Denver', state: 'CO', zip: 80210)
+bike_shop_2 = Merchant.create(name: "Meg's Bike Shop2", address: '123 Bike Rd.', city: 'Denver', state: 'CO', zip: 80203)
```
Comment (priyapower) [inline db/seeds.rb:16]: Since this was just from a lesson, we should probably reseed this file soon :)

File: spec/features/cart/checkout_spec.rb [hunk 2]
```diff
@@ -27,6 +29,54 @@
...
+      expect(page).to have_content("You must Register or Log In to finish checking out.")
+      expect(page).to have_link("Register")
+      expect(page).to have_link("Log In")
+    end
+
+    it "has a Register link which links to registration page" do
+      visit "/cart"
+
+      # Without the below, it was giving an Ambiguous Match error because there are two "Register"'s on the page if they're not logged in.  One in the nav, one in the message below.  When I put the ! in front of the within, with the hope that that means "not within", the test does pass ... but I am not sure it's actually working, or if the test just doesn't understand what that is trying to do and is skipping it entirely.
+      # The reason I think it's working is because line 50 *is* passing, which wouldn't make any sense if line 47 wasn't working.  Something to think on...
```
Comment (priyapower) [inline spec/features/cart/checkout_spec.rb:45]: WHOA - this seems like a pretty big view issue - Would renaming one of the links a little differently help give our users easier access to register or is it literally just that there are 2 places to register on this one web page?

Comment (priyapower) [reply spec/features/cart/checkout_spec.rb:45]: I will probably pull this down and check OR do a code review with you to answer this question

File: spec/features/cart/checkout_spec.rb [hunk 3]
```diff
@@ -27,6 +29,54 @@
...
+      !within 'nav' do
```
Comment (priyapower) [inline spec/features/cart/checkout_spec.rb:46]: I do like your workaround - look for register OUTSIDE of nav!

File: spec/features/cart/checkout_spec.rb [hunk 4]
```diff
@@ -27,6 +29,54 @@
...
+
+    it "has a Log In link which links to registration page" do
+      visit "/cart"
+
+      click_on "Log In"
+
+      expect(current_path).to eq("/login")
+    end
+
+    it "does not prompt useres to login before checking out" do
```
Comment (priyapower) [inline spec/features/cart/checkout_spec.rb:61]: typo = "users". No big deal, but thought I would mention

Comment (priyapower) [inline spec/features/cart/checkout_spec.rb:61]: Excellent sad path test! If I'm a logged in user, I SHOULDN'T be asked to login again! What an annoying user experience that would have been!

File: spec/features/items/index_spec.rb [hunk 5]
```diff
@@ -58,5 +58,120 @@
...
+        @okay_candy = @meg.items.create(name: "Okay candy", description: "You can eat it!", price: 3, image: "https://images-na.ssl-images-amazon.com/images/I/610ZOacuGPL._SL1050_.jpg", inventory: 100)
+        @bad_paper = @meg.items.create(name: "Bad Paper", description: "Bad for writing on!", price: 20, image: "https://cdn.vertex42.com/WordTemplates/images/printable-lined-paper-wide-ruled.png", inventory: 3)
+        @bad_pencil = @meg.items.create(name: "Bad Pencil", description: "You cannot write on paper with it!", price: 2, image: "https://images-na.ssl-images-amazon.com/images/I/31BlVr01izL._SX425_.jpg", inventory: 100)
+        @bad_candy = @meg.items.create(name: "Bad candy", description: "You cannot eat it!", price: 3, image: "https://images-na.ssl-images-amazon.com/images/I/610ZOacuGPL._SL1050_.jpg", inventory: 100)
+        @very_bad_paper = @meg.items.create(name: "Very Bad Paper", description: "Horrible for writing on!", price: 20, image: "https://cdn.vertex42.com/WordTemplates/images/printable-lined-paper-wide-ruled.png", inventory: 3)
```
Comment (priyapower) [inline spec/features/items/index_spec.rb:73]: I need to know about this very bad paper ;)

File: spec/features/items/index_spec.rb [hunk 6]
```diff
@@ -58,5 +58,120 @@
...
+        @very_bad_pencil = @meg.items.create(name: "Very Bad Pencil", description: "You get hurt by it!", price: 2, image: "https://images-na.ssl-images-amazon.com/images/I/31BlVr01izL._SX425_.jpg", inventory: 100)
+
+        11.times do
```
Comment (priyapower) [inline spec/features/items/index_spec.rb:76]: This is soooo much smarter. I need to update some of my tests like this. Maybe on Wednesday we can assign sections of refactoring: like I can take items/merchants, you take user/cart, etc, etc... Nothing set, just an idea

File: spec/features/items/index_spec.rb [hunk 7]
```diff
@@ -58,5 +58,120 @@
...
+        fill_in :city, with: city
+        fill_in :state, with: state
+        fill_in :zip, with: zip
+
+        click_button "Create Order"
+
+        visit '/items'
+      end
+
+      it "can show 5 most and least popular items and quantity bought" do
```
Comment (priyapower) [inline spec/features/items/index_spec.rb:144]: Is this US18? Did this come in on accident?

**Review by priyapower — APPROVED** [review]
This looks awesome! Some minor comments here and there and a typo I caught that doesn't affect the code :)

File: spec/features/cart/checkout_spec.rb (hunk 3 above)
Comment (LHJE) [reply spec/features/cart/checkout_spec.rb:46]: Thank you!  I was just trying something for now.  It'll have to be updated later to be more specific, but I don't quite have the parlance for how that should work.  I'm sure we'll find a better way!

File: spec/features/cart/checkout_spec.rb (hunk 4 above)
Comment (LHJE) [reply spec/features/cart/checkout_spec.rb:61]: I'll note it for the PR for US26!

File: spec/features/items/index_spec.rb (hunk 5 above)
Comment (LHJE) [reply spec/features/items/index_spec.rb:73]: Lol it's super bad

File: spec/features/items/index_spec.rb (hunk 6 above)
Comment (LHJE) [reply spec/features/items/index_spec.rb:76]: Yeah!  Glad we were able to implement it in the next PR!

File: spec/features/items/index_spec.rb (hunk 7 above)
Comment (LHJE) [reply spec/features/items/index_spec.rb:144]: Yeah, this must be a part of that.  I'll note it so when we merge Everything we're all working on into Dev that we make sure there isn't duplication on this.

**Pull request merged**